### PR TITLE
add support for nv-ipam deployment

### PR DESCRIPTION
**Pull request by adrianchiris**
Add support for nv-ipam deployment](https://github.com/Mellanox/network-operator/pull/502/commits/bbb23684a622a4d8a3e3fc21ed7955e229871db4) 

- Extend nicclusterpolicy with a new entry
  to enable deployment of nv-ipam CNI
- Add new State for nv-ipam deployment
- Add manifests for nv-ipam CNI deployment
- Update Helm chart and Readme

### CODE REVIEW

File: manifests/state-nv-ipam-cni/030-controller-config.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: nv-ipam-controller-config
+  namespace: {{ .RuntimeSpec.Namespace }}
+data:
+  controller-conf: |
```
Comment (moshe010) [inline manifests/state-nv-ipam-cni/030-controller-config.yaml:20]: in Yury code it is config https://github.com/Mellanox/nvidia-k8s-ipam/pull/5/files#diff-5c4df879eb29e6a4f6acc02753175b229a48f72c6b725a75cc00e9ab46bc33a2R29

File: config/rbac/role.yaml [hunk 2]
```diff
@@ -85,7 +85,6 @@ rules:
   - deployments
   - replicasets
   - statefulsets
-  - controllerrevisions
```
Comment (ykulazhenkov) [inline config/rbac/role.yaml:88]: Do we need this for upgradeController?

Comment (adrianchiris) [reply config/rbac/role.yaml:88]: i think its missing somewhere in annotation, i didnt actively remove it, in any case lemme revet this one

Comment (rollandf) [reply config/rbac/role.yaml]: Fixed in #505

**Review by e0ne — CHANGES_REQUESTED** [review]
Let's rebase this PR on top of #489 and merge it after #489

**Comment by adrianchiris** [comment]
@e0ne rebased, PTAL

File: deployment/network-operator/values.yaml [hunk 3]
```diff
@@ -225,6 +225,15 @@ ibKubernetes:
   pKeyGUIDPoolRangeEnd: "02:FF:FF:FF:FF:FF:FF:FF"
   ufmSecret: # specify the secret name here
 
+nvIpam:
+  deploy: false
+  image: nvidia-k8s-ipam
+  repository: ghcr.io/mellanox
+  version: latest
```
Comment (e0ne) [inline deployment/network-operator/values.yaml:232]: do we really whant to use `latest` tag? we don't use it with other components

File: manifests/state-nv-ipam-cni/040-nv-ipam-controller.yaml [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  template:
+    metadata:
+      labels:
+        component: nv-ipam-controller
+        app: nv-ipam
+        name: nv-ipam-controller
+    spec:
+      priorityClassName: system-cluster-critical
+      serviceAccountName: nv-ipam-controller
+      # required to be scheduled on a linux node with node-role.kubernetes.io/master label and
```
Comment (e0ne) [inline manifests/state-nv-ipam-cni/040-nv-ipam-controller.yaml:43]: Please, add `"beta.kubernetes.io/os": "linux"` selector or delete this comment

**Comment by rollandf** [comment]
Do we need to also update here?
https://github.com/Mellanox/network-operator/blob/master/hack/release.go#L46

**Review by e0ne — CHANGES_REQUESTED** [review]

File: deployment/network-operator/values.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (adrianchiris) [reply deployment/network-operator/values.yaml:232]: we do :), will update to v0.0.2

Comment (adrianchiris) [reply deployment/network-operator/values.yaml:232]: we do :), will update to v0.0.2

File: manifests/state-nv-ipam-cni/040-nv-ipam-controller.yaml (hunk 4 above)
Comment (adrianchiris) [reply manifests/state-nv-ipam-cni/040-nv-ipam-controller.yaml:43]: will delete comment

File: manifests/state-nv-ipam-cni/040-nv-ipam-controller.yaml [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      {{- end }}
+      {{- end }}
+      containers:
+        - name: nv-ipam-controller
+          image: {{ .CrSpec.Repository }}/{{ .CrSpec.Image }}:{{ .CrSpec.Version }}
+          imagePullPolicy: IfNotPresent
+          command: ["/ipam-controller"]
+          args:
+            - --config-name=nvidia-k8s-ipam-config
+            - --config-namespace=$(POD_NAMESPACE)
```
[the remark this answers is not in the document]
Comment (adrianchiris) [reply manifests/state-nv-ipam-cni/040-nv-ipam-controller.yaml:85]: Thats how it defined in [1]. I prefer to keep it close to the original definition.

[1] https://github.com/Mellanox/nvidia-k8s-ipam/blob/main/deploy/nv-ipam.yaml

**Comment by adrianchiris** [comment]
@e0ne @rollandf LMK if you are happy with the PR so we can merge.

**Review by e0ne — APPROVED** [review]
LGTM. Thanks for addressing my comments

**Review by e0ne — APPROVED** [review]
LGTM. Thanks for addressing my comments

**Comment by rollandf** [comment]
LGTM

**Pull request merged**